### PR TITLE
Drawing cleanup

### PR DESCRIPTION
**Pull request by JoshKarpel**
Wide variety of internal and external improvements to the `drawing` submodule:

- change variable names throughout to be more descriptive and drop unnecessary type prefixes
- break up the large `update` functions to make the classes more flexible
- add the test scripts under `dev/`, and update them to use the new API from the previous bullet

**This PR has breaking changes for all existing drawing code!** Please inspect carefully...

Since there are a lot of individual commits in here that don't represent good final states, you should do a "squash merge" (https://help.github.com/en/articles/about-pull-request-merges#squash-and-merge-your-pull-request-commits).

### CODE REVIEW

**Comment by ecpoppenheimer** [comment]
Two comments:
1) I am fine using fstrings, but before I do that I guess I am going to have to figure out how to get my python environment set up better, because right now I am still on 3.5.2.  So it will take a few days for me to get that sorted out.

2) You changed the shape requirements on the ray tensor to requiring the ray tensor to have exactly 5 elements in its second dimension, instead of at least 5.  The >= there is important, because most ray tensors passed to the ray drawer will actually have 6 elements in the second dimension.  The reason for this is that the ray tracer appends a sixth element to all of the ray sets it outputs which represents the index of the parent ray that generated each of the output rays.  This in turn is useful for sorting the output rays, or comparing them to some property of the input rays.  This is necessary because the ray tracer might reorder the output rays, especially if some of them undergo fewer interactions than others.

An example of when I have used this: setting up optimization problems with an extended source.  If you have a point source, the goal is to get all of the output rays to hit the focal plane at zero.  But if you have an extended source, and you want the optimization to care about things like coma, you don't want the rays to all hit the focal plane at zero, you want each output ray to hit the focal plane at a location that depends on where along the extended source that ray originated from.  To achieve this, you can use that sixth element (which is just an index that points to input rays) to figure out where that output ray should be hitting the focal plane.

TLDR: RayDrawer may get ray tensors that are (n, 5) or (n, 6), but it doesn't need anything beyond 5 elements to draw the rays.  The sixth element won't be used by all users, and I don't want to make them have to slice it off manually.

**Comment by JoshKarpel** [comment]
I pushed some new commits that revert the stricter length requirements

**Comment by ecpoppenheimer** [comment]
I have looked over your changes to drawing.py.  Most of it looks good, and I am close to ready to merge, though I do have one small thing I want to change.  Is the proper workflow for me to tell you what I want you to change, or should I squash and merge, and then make my tweaks, or can I jump onto your branch, make the changes, and then merge?

The change is around the maybe_draw_canvas stuff.  The automatic canvas redraw stuff altogether is probably not necessary.  The way that all matplotlib drawing tools work is that you make whatever changes you want to the artist, and then you redraw the canvas.  I included the auto_redraw_canvas option for the benefit of being lazy, and then have never used it because redrawing the canvas can be a slow operation and I want to do it as rarely as possible.  I have used tfrt to model a frustrated TIR waveguide which took over a second to draw the canvas.  And when I am tweaking an optimizer, I will usually start redrawing the canvas after every update, but once I am confident that the optimizer is working smoothly, I will drop to redrawing the canvas every 10 updates or less which can bring a speedup of 2x or more.

So we could completely eliminate the auto redraw stuff and just make the drawing API more like the matplotlib API.  Or we could keep it as a convenience.  But if we keep it either:
1) The self._maybe_draw_canvas() call should be removed from ArcDrawer.arrow_count.setter, style.setter, and color.setter (similar in SegmentDrawer) because those setters won't actually touch the matplotlib artists, and so can't have an effect on the canvas that calls for a redraw.
2) Cause the above setters to also call draw(), so that they actually effect a change on the matplotlib artists.

I am leaning toward removing all of the auto redraw stuff.  What do you think?  Maybe we can have a module level function that redraws the canvas, just as a minor convenience to the user in case they don't know how to use matplotlib?

--------------------------

Also, if we have a setter for norm arrow count, we should probably also have one for norm arrow length.  I have never needed to change the arrow count, but the arrow length is important to change, since I have written scripts that operate on very different length scales.  My frustrated waveguide was 3 mm x 30 cm, and had arcs (representing dots of printed transparent ink) < .1 mm radius.  For that one, checking surface orientation required very different scales for the arrows.

**Comment by JoshKarpel** [comment]
Makes sense to me, I'll remove auto redraw and add an arrow length setter on my end.
